src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp

```diff
@@ -332,5 +335,53 @@ void MetadataStorageFromPlainObjectStorageUnlinkMetadataFileOperation::undo(std:
             path);
     }
 }
+MetadataStorageFromPlainObjectStorageCopyFileOperation::MetadataStorageFromPlainObjectStorageCopyFileOperation(
+    std::filesystem::path path_from_,
+    std::filesystem::path path_to_,
+    InMemoryDirectoryPathMap & path_map_,
+    ObjectStoragePtr object_storage_)
+    : path_from(path_from_)
+    , remote_path_from(object_storage_->generateObjectKeyForPath(path_from_, std::nullopt).serialize())
+    , path_to(path_to_)
+    , remote_path_to(object_storage_->generateObjectKeyForPath(path_to_, std::nullopt).serialize())
+    , path_map(path_map_)
+    , object_storage(object_storage_)
+{
+}
+
+void MetadataStorageFromPlainObjectStorageCopyFileOperation::execute(std::unique_lock<SharedMutex> & /*metadata_lock*/)
+{
+    LOG_TEST(getLogger("MetadataStorageFromPlainObjectStorageCopyFileOperation"), "Copying file from '{}' to '{}'", path_from, path_to);
+
+    const auto directory_from = path_from.parent_path();
+    if (!path_map.existsLocalPath(directory_from))
```

Comment:
Maybe we should check for a source file, not a directory?

Reply:
done